src/Caching/Hybrid/src/Internal/DefaultHybridCache.cs

```diff
@@ -4,59 +4,159 @@
 using System.Linq;
+using System.Runtime.CompilerServices;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Caching.Memory;
 using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Options;
 
 namespace Microsoft.Extensions.Caching.Hybrid.Internal;
 
 /// <summary>
 /// The inbuilt ASP.NET implementation of <see cref="HybridCache"/>.
 /// </summary>
-internal sealed class DefaultHybridCache : HybridCache
+internal sealed partial class DefaultHybridCache : HybridCache
 {
-    private readonly IDistributedCache _backendCache;
-    private readonly IServiceProvider _services;
+    private readonly IDistributedCache? _backendCache;
+    private readonly IMemoryCache _localCache;
+    private readonly IServiceProvider _services; // we can't resolve per-type serializers until we see each T
+    private readonly IHybridCacheSerializerFactory[] _serializerFactories;
     private readonly HybridCacheOptions _options;
+    private readonly ILogger? _logger;
+    private readonly CacheFeatures _features; // used to avoid constant type-testing
 
-    public DefaultHybridCache(IOptions<HybridCacheOptions> options, IDistributedCache backendCache, IServiceProvider services)
+    private readonly HybridCacheEntryFlags _hardFlags; // *always* present (for example, because no L2)
+    private readonly HybridCacheEntryFlags _defaultFlags; // note this already includes hardFlags
+    private readonly TimeSpan _defaultExpiration;
+    private readonly TimeSpan _defaultLocalCacheExpiration;
+
+    private readonly DistributedCacheEntryOptions _defaultDistributedCacheExpiration;
+
+    [Flags]
+    internal enum CacheFeatures
+    {
+        None = 0,
+        BackendCache = 1 << 0,
+        BackendBuffers = 1 << 1,
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private CacheFeatures GetFeatures(CacheFeatures mask) => _features & mask;
+
+    internal CacheFeatures GetFeatures() => _features;
+
+    // used to restrict features in test suite
+    internal void DebugRemoveFeatures(CacheFeatures features) => Unsafe.AsRef(in _features) &= ~features;
+
+    public DefaultHybridCache(IOptions<HybridCacheOptions> options, IServiceProvider services)
     {
-        _backendCache = backendCache ?? throw new ArgumentNullException(nameof(backendCache));
         _services = services ?? throw new ArgumentNullException(nameof(services));
+        _localCache = services.GetRequiredService<IMemoryCache>();
         _options = options.Value;
+        _logger = services.GetService<ILoggerFactory>()?.CreateLogger(typeof(HybridCache)); // note optional
+
+        _backendCache = services.GetService<IDistributedCache>(); // note optional
+
+        // ignore L2 if it is really just the same L1, wrapped
+        // (note not just an "is" test; if someone has a custom subclass, who knows what it does?)
```

Comment:
Do we expect custom subclasses of `MemoryDistributedCache` and `MemoryCache` to be common here?

Reply:
nope, but they're not `sealed` so I can't rule it out